### PR TITLE
add dfhack string functions and tests

### PR DESCRIPTION
**Pull request by myk002**
step 2 of #1888

this implementation of `string:split` is based on `utils.split_string()` but adds infinite loop protection and the ability to treat the delimiter as a literal instead of as a pattern.

- added string:split
- added string:trim
- added a default value for wrap width
- added function comments
- added tests for all string functions (the tests for string:split will be commented out until we remove the competing implementation in gui/load-screen

### CODE REVIEW

File: test/library/string.lua [hunk 1]
```diff
@@ -18,6 +21,41 @@ function test.endswith()
...
+    -- we don't actually care what this returns, just that it does return
+    expect.true_(('hello world'):split('.*'), 'ensure no infinite loop')
+
+    expect.table_eq({'hello ', ' world'}, ('hello , world'):split(','),
+                    'ensure spaces are kept when they are not the delimiter')
+    expect.table_eq({'hello'}, ('hello'):split(), 'no delimiter')
+end
+]]
+
+function test.trim()
```
[the remark this answers is not in the document]
Comment (myk002) [reply test/library/string.lua:48]: All test code in this PR is new

[the remark this answers is not in the document]
Comment (lethosor) [reply test/library/string.lua:48]: Oops, I meant to comment on the addition of `string.trim()`. I saw a similar C++ function in uicommon.h, but not one in Lua

[the remark this answers is not in the document]
Comment (myk002) [reply test/library/string.lua:48]: Oh, yes, that function (and it's tests) are new. Are you thinking it should be passed through to the C++ implementation?

[the remark this answers is not in the document]
Comment (lethosor) [reply test/library/string.lua:48]: Nope, just wanted to confirm in case #1888 needed to be updated

**Review by lethosor — APPROVED** [review]
Looks good! Good catch on the issue with empty delimeters.

**Comment by myk002** [comment]
> Looks good! Good catch on the issue with empty delimeters.

Also the less obvious case of patterns that match 0 characters, like ',*'

**Pull request merged**